feat(acme): implement resources and data sources for ACME accounts

ZauberNerd opened the pull request:
This PR implements the API for managing a Proxmox Cluster's ACME accounts and the related data sources and resources.

### Contributor's Note
<!--- 
Please mark the following items with an [x] if they apply to your PR.
Leave the [ ] if they are not applicable, or if you have not completed the item.
--->
- [ ] I have added / updated documentation in `/docs` for any user-facing features or additions.
- [ ] I have added / updated acceptance tests in `/fwprovider/tests` for any new or updated resources / data sources.
- [ ] I have ran `make example` to verify that the change works as expected.

<!---
You can find more information about coding conventions and local testing in the [CONTRIBUTING.md](https://github.com/bpg/terraform-provider-proxmox/blob/main/CONTRIBUTING.md) file.

If you are unsure how to run `make example`, see [Deploying the example resources](https://github.com/bpg/terraform-provider-proxmox?tab=readme-ov-file#deploying-the-example-resources) section in README.
--->

<!--
*IF* your code contains breaking changes make sure to add `!` to the end of commit type, e.g.:
```
    feat(vm)!: add support for new feature 
```
Also, uncomment the section just below, and add a description of the breaking change. 
--->

<!---
#### ⚠ BREAKING CHANGES

>>> Put your description here <<<
--->

### Proof of Work
<!--- 
Please add screenshots, logs, or other relevant information that demonstrates the change works as expected.
--->

<!--- Please keep this note for the community --->
### Community Note

- Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
- Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request
<!--- Thank you for keeping this note for the community --->

<!--- If your PR fully resolves and should automatically close the linked issue, use Closes. Otherwise, use Relates --->

<!--- Release note for [CHANGELOG](https://github.com/bpg/terraform-provider-proxmox/blob/main/CHANGELOG.md) will be created automatically using the PR's title, update it accordingly. --->

ZauberNerd commented on fwprovider/acme/resource_acme_account.go:
```diff
@@ -0,0 +1,258 @@
+	state.Directory = types.StringValue(account.Directory)
+	state.TOS = types.StringValue(account.TOS)
+	// XXX account.Location?
+	// XXX account.Account?
```
See account above about letsencrypt and the `data` value. `GET`ting a specific ACME account also returns the `account` object and a `location` url, which in the case of letsencrypt is a url to the letsencrypt account.
https://pve.proxmox.com/pve-docs/api-viewer/#/cluster/acme/account/{name}

I guess these should go into the state as `computed` properties?

ZauberNerd replied:
In contrast to the data source (https://github.com/bpg/terraform-provider-proxmox/pull/1455#discussion_r1701744820) I think for the resource I need to extract the first email address from `data.contact` and put that into `state.Contact`, right?
Because in the `POST` https://pve.proxmox.com/pve-docs/api-viewer/#/cluster/acme/account I'm providing `contact`, `directory`, `eab-hmac-key`, `eab-kid`, `name` and `tos_url` 
but `GET`ting an account (https://pve.proxmox.com/pve-docs/api-viewer/#/cluster/acme/account/{name}) returns `account`, `directory`, `location` and `tos` so that should be mapped to the state, right?

ZauberNerd replied:
I updated the PR to take the first email address from `contact` and write that into the state.

ZauberNerd commented on fwprovider/acme/resource_acme_account.go:
```diff
@@ -0,0 +1,258 @@
+
+	updateRequest := &account.ACMEAccountUpdateRequestBody{}
+	updateRequest.Contact = plan.Contact.ValueString()
+	updateRequest.Name = plan.Name.ValueString() // XXX name can't be changed?
```
`PUT` (https://pve.proxmox.com/pve-docs/api-viewer/#/cluster/acme/account/{name}) has two parameters, `contact` and `name`. But it looks like only `contact` can be changed, when trying to change `name` it says either that:
- the account was not found (makes sense, because `name` is in the path)
- or, if I use the old name in the path and the new name as body parameter, it says duplicate parameter.

I'm not sure how to deal with that in this resource.

ZauberNerd commented on fwprovider/acme/resource_acme_account.go:
```diff
@@ -0,0 +1,258 @@
+	updateRequest.Contact = plan.Contact.ValueString()
+	updateRequest.Name = plan.Name.ValueString() // XXX name can't be changed?
+
+	err := r.client.Update(ctx, plan.Name.ValueString(), updateRequest)
```
See above, it probably makes sense to use the name from state here instead of plan?

ZauberNerd replied:
After thinking about this for a while, I reverted this change, because I think changing the name should fail instead of silently succeed.

ZauberNerd commented on fwprovider/acme/resource_acme_account.go:
```diff
@@ -0,0 +1,258 @@
+
+	plan.Directory = types.StringValue(account.Directory)
+	plan.TOS = types.StringValue(account.TOS)
+	// XXX account.Location?
+	// XXX account.Account?
```
See above about location and account

ZauberNerd commented on proxmox/cluster/acme/account/client.go:
```diff
@@ -0,0 +1,31 @@
+	"fmt"
+
+	"github.com/bpg/terraform-provider-proxmox/proxmox/api"
+	"github.com/bpg/terraform-provider-proxmox/proxmox/nodes/tasks"
```
nitpick: `tasks` currently live under `nodes`, but it seems like tasks are a cluster-wide resource.

ZauberNerd commented on fwprovider/acme/datasource_acme_account.go:
```diff
@@ -0,0 +1,150 @@
+	// Name is the ACME account config file name.
+	Name types.String `tfsdk:"name"`
+	// Account is the ACME account information.
+	// Account types.Map `tfsdk:"account"` // XXX
```
I accessed the `https://acme-staging-v02.api.letsencrypt.org/acme/acct/xxxx` API directly and it looks like the `account` field is just passed through from the letsencrypt API and at least `status` and `contact` are specced: https://datatracker.ietf.org/doc/html/rfc8555/#section-7.1.2
The Proxmox web UI code then also just takes the first element of `data.account` as the email address, so I assume we could do that as well.

ZauberNerd replied:
For the data source I am just passing through the `contact: []string` array into the state. Or should I orient myself on the UI and take the first element of `contact`?

bpg reviewed (commented):
Hi @ZauberNerd 👋🏼 

Thanks so much for contributing this! I think I've replied to all of your questions, but please ping me if you want to discuss something else.

Could you please also add the new resource/datasources to the `/tools/tools.go` list, then run `make docs` and commit the result? This will update the generated provider's docs.

I'd also love to see some acceptance tests for the resource, but I'll probably add them myself after you finalize your code. 
That would help me with the review as well :)

ZauberNerd commented on fwprovider/acme/resource_acme_account.go:
```diff
@@ -0,0 +1,258 @@
+	// Key Identifier for External Account Binding.
+	EABKID types.String `tfsdk:"eab_kid"`
+	// URL of CA TermsOfService - setting this indicates agreement.
+	TOS types.String `tfsdk:"tos"`
```
I'll add the meta data source in another PR. For now this resource is functional by providing the URLs manually.

ZauberNerd commented:
@bpg I've updated the PR and added a few comments.
Could you please take another look?

> I'd also love to see some acceptance tests for the resource, but I'll probably add them myself after you finalize your code.
> That would help me with the review as well :)

I'm happy with adding acceptance tests myself, let me know which you prefer.

bpg commented:
@ZauberNerd I started looking into adding acceptance tests and immediately hit a road bump. For some reason, the `POST /api2/json/cluster/acme/account/` API can't work via API Token authentication and requires the user to be authenticated with local `root@pam`:
```
error creating ACME account: received an HTTP 403 response - Reason:
        Permission check failed (user != root@pam)
```

I don't really like switching all acceptance tests to use root account authentication, so I'll need to update the test harness to use different authentication types for different tests.

However, I also don't want to hold this PR up much longer. I think it is in a good state. I'll run a few more manual tests tomorrow and approve it soon.

bpg reviewed (approved):
Awesome work, thanks again for adding this feature!

LGTM! 🚀

bpg commented:
@all-contributors please add @ZauberNerd for code, idea
